### PR TITLE
Avoid to bypass `try_new/new()` to build plan directly and cleanup filter

### PR DESCRIPTION
**Pull request by jackwener**
# Which issue does this PR close?

Closes #.

# Rationale for this change

# What changes are included in this PR?

- add `#[non_exhaustive]` for LogicalPlan to avoid to bypass `try_new/new()` to build plan directly and 
- cleanup filter `predicate() input()`

# Are these changes tested?

# Are there any user-facing changes?

### CODE REVIEW

File: datafusion/optimizer/src/optimizer.rs [hunk 1]
```diff
@@ -502,26 +502,27 @@ mod tests {
...
+        let input = Arc::new(test_table_scan()?);
         let input_schema = input.schema().clone();
 
-        let plan = LogicalPlan::Projection(Projection {
-            expr: vec![col("a"), col("b"), col("c")],
+        let plan = LogicalPlan::Projection(Projection::try_new_with_schema(
+            vec![col("a"), col("b"), col("c")],
             input,
-            schema: add_metadata_to_fields(input_schema.as_ref()),
-        });
```
Comment (jackwener) [inline datafusion/optimizer/src/optimizer.rs:518]: original we can construct `Project` bypass the `try_new()`

File: datafusion/optimizer/src/single_distinct_to_groupby.rs [hunk 2]
```diff
@@ -98,6 +98,7 @@ impl OptimizerRule for SingleDistinctToGroupBy {
                 aggr_expr,
                 schema,
                 group_expr,
+                ..
```
Comment (alamb) [inline datafusion/optimizer/src/single_distinct_to_groupby.rs:101]: this is unfortunate (previously when new fields were added, the compiler would tell you all locations that might need to be changed. I realize this is a consequence of marking the structs `non_exhaustive`

**Review by alamb — APPROVED** [review]
Thank you @jackwener 
I really like how this make the codebase more uniform

It appears this PR now has some conflicts

**Comment by alamb** [comment]
Thanks @jackwener

**Pull request merged**